README.md
# PassCrack 
A personal project which started as an assignment for my "Computer Security" course, developed for almost two semesters.
PassCrack is a powerful passsword cracker written in C/C++ that uses Rainbow Tables and (currently) works for passwords hashed with the _Blake_ algorithm.
It is a molular and configurable enough to demonstrate the _Time-Memory Tradeoff_, originally analyzed by [Phillipe Oechslin](http://lasec.epfl.ch/~oechslin/publications/crypto03.pdf) and to support other hashing algorithms, password alphabets, reductions functions etc.



## Code Layout
File | Purpose
-----|--------
README.txt | The assignment's description, written in greek. Original Documentation.
PassCrack | The application binary, current release. (Compsec16)
main.cpp | The code for the main thread. Parses command line arguments, orchestrates spawned threads, defines some globals and contains the code-backbone
rainbowfuncts.cpp | Implementations of the general-purpose functions used.
generate.cpp | Defines functions regarding table generation.
test.cpp | Testing module
blake.cpp, blake.hpp | Implementation and Interface files for the BLAKE256 hash function used. Based on the [standard](https://131002.net/blake/#dl).
tablemodel.txt, chainmodel.txt | Table and chain demos
makefile | Build script
generator-script.bash | A bash script to automate back2back runs for small tables to avoid memory overuse.
changelog.txt | A diary of major fixes between versions.
tablemerger.cpp | Just a piece of code to merge small tables together into a big one with unique endpoints.

__NOTE:__ This programm uses modern libraries, part of the C++11 standard, which the compiler used must support.



## Design & Operations
The operations carried out by this programm can be classified in 2 modes:

1. OFFLINE mode, where the rainbow table is generated or imported and execution parameters are set.
2. ONLINE mode, where given a hashed password, a "Reduce-Lookup-Hash" loop is initiated.

Supposing an environment where the password must be "reversed" (from the hash) in a short period of time, with the password being refreshed and rehashed after the end of the period, an event to halt the programm is required. Consequently, a SIGTSTOP signal (Ctr+Z) stops the loop and a "new hash" prompt is spawned.

Similarly, to terminate the programm a SIGINT (Ctr+C) must be signalled.

### Reduction Functions
The purpose of these routines is to map -as "one-way" as possible"- a 32byte hash to a 6character password from an alphabet of 64 valid characters (!,@, 1,2,...,9, A,B,...,Z, a,b,...,z). Therefore (pseudo-)randomness and entropy are the drivers behind each technique. 
The current version include a set reduction-function-families, but decoupling the _password alphabet_ extra families can be implemented, and instantly used. Each family defines several reduction functions based on the same tecnhique, slightly altered depending on each function's index. The techniques per family are the following:

* __Set #1: "Semi ASCII"__ [32 indices]

  All 32 functions choose 6 bytes from the hash. 
  * The first one picks bytes 0,1,...,5.
  * The second one picks bytes 1,2,...,6,
  
  and so it goes until the 33rd byte is requested where the 1st is given instead. Continuing with a mod32 arithmetic, the number of genuinely different functions is limited to 32.
  For each one of these 6 bytes, after zero-ing out the msb to match the 7bit ASCII code, the values are matched to a character like this:
    
    00 ...... 2F | 30 ...... 39 | 3A ... 40 | 41 ...... 5A | 5B ... 60 | 60 ...... 7A | 7B ... 7F
    -------------|--------------|-----------|--------------|-----------|--------------|----------
    !       | 0, 1, ..., 9  |    @      | A, B, ...,Z  |     !     |  a, b, ...,z  |     @


* __Set #2: "High Kolmogorov Complexity"__ [992 indices]

  All functions choose 6 bytes from the hash and map them using mod64 arithmetic as above. Every function belongs to only one of the 31 "classes". Each class includes the 32 genuinely different functions, all of which start the selection from the same byte. e.g.:
  * Functions of the 1st class start with byte 0.
  * Functions of the 2nd class start with byte 1 etc.
  Then, the next byte for each is selected adding the function's intra-class index (1-32) to the selected byte's number, cicrulating around 32. e.g.:
  * Function #28 of class 2 selects bytes 28, 30, 0, 2, 4, 6.
  
  
* __Set #3: "Mix & Match"__ [32 indices]

  Driven by simplicity and calculation time (in CPU cycles), this technique is based on a mod64 logic, salted by some "magic numbers".
  
  
* __Set #4: "PRNG/Uniform indices"__ [infinite indices]

  Following the common "PRNG & hash" scheme, a generator is seeded with the function's index. After 6 calls, the generated numbers are the selectors for 6 bytes from the hash, the values of which are mapped to characters from the alphabet using mod64 arithmetic.
  
  
* __Set #5: "PRNG/Uniform characters"__ [infinite indices]

  Following the common "PRNG & hash" scheme, a generator is seeded with the value of a byte. The function's index is the selector for this byte. After 6 calls, the generated numbers are mapped to password characters from the alphabet.
  
  
* __Set 36: "Human-like"__ [for future implementation]

  A family of functions that will meet the second Reduction Function Requirement (besides high entropy). That is, to output passwords that a human would probably select, or at least pronεξασφαλιζωounceable (thus memorable!).
  An idea to achieve the above, would be to map bytes against an alphabet of phonems/digramms/syllables that are most commonly used in most languages.
  Another idea is to inject dates (e.g. 1994) inside possible passwords mimicing the human behavior.
  Such behavior was not required for this assignment since the password was randomly chosen.
  
  

## Command Line Options
Make sure you have read and understood the Design Principles listed above, before proceeding.
```bash
  ./PassCrack [online|offline] [-uvcgqQ       ] [-m <chains>  ]
              [-h <load>     ] [-i <intable>  ] [-t <links>   ]
              [-p <threads>  ] [-o <outable>  ] [-l <tables>  ]
              [-r <redfunset>] [-s <outchain> ]
```

Where:
  * No option is mandatory
  * Options can be in any order, as long as the options requiring an argument are followed by this argument.
  * Specifically:

    * ```online```/```offline``` Mode of operation. Online mode requires table input/generation (default).
    * ```-u (usage)``` Print this guide.
    * ```-v (verbose)``` Print extra info during execution. Recommended.
    * ```-c (calculate)``` Calculate success probability for the given parameters using Oechslins formula.
    * ```-g (generator)``` Use the modern, more robust PRNG defined in the C++11 standard for the table generation.
    * ```-q (unique)``` Enable table post-generation processing to discard duplicate chains (with similar endpoints). Option only meaningful if the table is generated and not imported.
    * ```-Q (unique)``` Like ```-q``` but each chain's uniqueness is examined right after creation. Ensures table size equal to the requested (see ```-m``` option) with a cost of many CPU cycles. Option only valid if table is generated by a single thread (see ```-p``` option).
    * ```-r (reduction)``` Use ```<redfunset>``` family for the reduction functions. Defaut: ```3```.
    * ```-h (hashmap)``` Convert the table to a hash map for O(1) Lookup complexity. It is also possible to declare eac bucket's "load factor". Default: ```100.0```.
    * ```-p (parallel)``` Use ```<threads>``` threads to generate the table concurrently. Option only valid if table is not imported (see ```-i``` option). If combined with ```-s``` option, only the first thread's chains are printed for verification.
    * ```-i (input) ```/```-o (output)``` Input/output the table from/to the ```<intable>```/```<outtable>``` file. Files should have the format specified in the [model file](https://github.com/LAripping/PassCrack/blob/master/tablemodel.txt). Both absolute and relative paths are supported. The length of each Reduce-Lookup-Hash loop depends on the ```t``` parameter which should match the table's length, even if imported. Therefore ```-i``` option must be combined with the ```-m``` and ```-t``` ones.
    * ```-s (save) ``` Export the 10 first complete chains in the ```<outchain>``` file for debugging/verification. Not recommended due to the danger of creating massive files. The written file has a format similar to the [model](https://github.com/LAripping/PassCrack/blob/master/chainmodel.txt). Both absolute and relative paths are supported.
    * ```-m``` Explicit definiton of table parameters 'm' (total chains per table) and 't' (total chains per table). Defaut: ```m=100, t=32```.
    * ```-l``` Create ```<tables>``` tables in total. Default and currently only supported value: ```1```.
    


changelog.txt
ver1.0:
-Λαθος λογική, στο follow chain
-Μοναδική redfun

ver2.0:
+Σωστη λογική στο follow chain
+Συνολο απο redfuncs
-Λαθος στους υπολογισμους για τους χρόνους που εμφανίζονται (αρνητικοί)

ver2.1:
+Διορθώθηκε το λάθος στους αρνητικούς χρόνους, τώρα υπολογίζονται και τα "κρατούμενα"
-λανθασμένη λογική εφαρμογής μονο της πρώτης red-fun στην παραγωγή πινακα
-λανθασμένη λογική της εφαρμογής των red-fun με την σειρα στο ONLINE PART

ver3.0:
+Εφαρμογή όλων των red-funs με τη σειρά στην παραγωγή πίνακα
+Εφαρμογή των red-fun με αντίστροφη σειρά στο ONLINE PART απο την αρχή κάθε φορά
 και με την ορθή στην αναπαραγωγή της αλυσίδας αν βρεθεί HIT
+Ενσωμάτωση υπολογιστή πιθανότητας επιτυχίας σύμφωνα με την φόρμουλα του Oechslin

ver4.0:
+Δυνατότητα παραλληλοποίησης στην παραγωγή πίνακα
Τα αρχεία Leo/Desktop/Blake[Table|Chain]_2Mx500.txt εχουν παραχθεί απο αυτή την εκδοση για δοκιμές
(για αυτό η συγκεκριμένη εκδοση δεν αναπτύχθηκε περαιτέρω)

ver4.6:
+Διάσπαση της λειτουργικότητας παραγωγής πίνακα σε ξεχωριστό αρχείο για καλύτερη συντήρηση
+Βημα μετα-επεξεργασίας που διαγράφει όλες τις αλυσίδες με κοινά endpoints 
    ->δραματικά ποσοστά για το 1ο σύνολο red-funs (194/200.000 διαφορετικές!)
+Υλοποίηση 2ου, 3ου σύνολου απο red-funs

ver4.7
-Aντικατάσταση της rand/rand_r για τα startpoints κατα την δημιουργία πίνακα
 με την uniform_int_distribution με default (time/time+thread_id seed) generator
 απο την βιβλιοθήκη <random> του standard C++11.
-Υλοποίηση 4ου, 5ου συνόλου απο red-funs που χρησιμοποιούν την παραπάνω PRNG
    -> κανένα απο τα δύο ομως δεν παρουσιάζει καλύτερη εντροπία ή ταχύτητα απο το 3ο! 
-Προστασία αντικατάστασης υπαρχοντων αρχείων εξοδόυ
-Εξαντλητική απόρριψη+αναδημιουργία αλυσίδων μέχρι να μην υπάρχουν στον πίνακα.
-Μεταγλώττιση με -Ο3

ver4.8
-Μετατροπή του πίνακα σε Hashmap (std::unordered_map) κατα το Online Mode για ταχύτατες
 αναζητήσεις πολυπλοκότητας Ο(1) (στην μέση περίπτωση). Αποθηκευση ζευγων {endpoint,startpoint}. Υποστηρίζεται επίσης και η επιλογή του "load factor". 



rainbowfuncts.cpp
#include <iostream>
#include <fstream>

#include <list>

#include <cstdint>
#include <cstdlib>
#include <cmath>
#include <cstring>

#include <random>


#include "blake.hpp"




using namespace std;
typedef list<pair<char[7],char[7]>*> Table_t;


char alphabet[64] = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '!', '@', 		\
		   			 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm',	\
					 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',	\
					 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M',	\
					 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', };




extern void (*hashfun)( uint8_t *out, const uint8_t *in, uint64_t inlen );
extern void (*redfun) ( char* out, const uint8_t *in, int red_by );
extern bool pass_comparator(const pair<char[7],char[7]>* first, const pair<char[7],char[7]>* second);	

extern bool		v, savechain;



/*
 * Reduction-functions' set #1.
 *          "Simple but terribly biased"
 *
 * Determined by the index "red_by" of the function selected from the set
 * 6 bytes are chosen from the hash.
 *
 * First their msb is unset to match the old 7bit ASCII format.
 * Then each byte is mapped to a character from the password's alphabet 
 *
 * Redfun #1:  takes bytes 0-5 (red_by>=1)
 *  -\\-  #2:  takes bytes 1-6
 *  -\\-  #3:  takes bytes 2-7 
 * ...
 * Redfun #32  takes bytes 31-4
 *
 * After 32 modulo indexing is used so #33 chooses 0-5 and so on.
 *
 * In first rep Rt is used, in second Rt-1,H,Rt, then Rt-2,H,Rt-1,H,Rt, and so it goes
 * until all function are used, then the attack fails
 *
 */
void red_functs_set1( char* out, const uint8_t *in, int red_by ){													
	int 	o_idx=0, i_idx;
	uint8_t byte, mask;
	
	for(int i=0 ; i<6 ; i++){																
		i_idx = (i + red_by - 1) % 32;						// Select input's byte to map
		byte  = in[ i_idx ];									
		
		mask = 0x7f;										// Create byte 01111111 to perform bitwise AND with hash's byte...
		byte &= mask;										// To unset hash-byte's 8th lefmost (msb) bit.
		
						
		if( ( 0x2f< byte && byte<=0x39 )					// Map them to a specific character in password's alphabet
		||  ( 0x41<=byte && byte <0x5b ) 
		||  ( 0x61<=byte && byte <0x7b ) )						
			out[o_idx] = (char) byte;						// This character is usually the one whose ASCII code 
		                                                    // equals the byte's value
		else if((byte>=0x3a && byte<=0x41)  
			||  (byte>=0x7b && byte<=0x7f) )
			out[o_idx] = '@';
		
		else if((byte>=0x00 && byte<=0x2f)
			||	(byte>=0x5b && byte<=0x60) )
			out[o_idx] = '!';
															// In case i forgot a case!
		else												// Control should never come here!
			cerr << "Case not supported!" << (int)byte << endl;	
				
		o_idx++;
				
	}
	out[o_idx] = '\0';										// NULL terminate it to process as C-string
}




/*
 * Reduction-functions' set #2.
 *          "High Kolmogorov Complexity"
 *
 * 1024 original red-funs
 *
 * Each one selects 6 bytes of the hash and MODs them with 64 to map it
 * to an alphabet letter.
 *
 * Discreet red-funs are divided into 31 "klasses" subsets, each klass 
 * with 32 red-funs defining the starting byte to be selected 
 *   1st starts with byte[0]
 *   2nd starts with byte[1]
 *   ...
 *
 * Then, the next byte is the previous +klass circularly around 
 * the 32 of the hash.
 * E.g. for klass=2, red-fun #28 selects bytes[28-30-0-2-4-6]
 *     
 *
 */
void red_functs_set2( char* out, const uint8_t *in, int red_by ){				
    
    int klass = (red_by/32 + 1);                        // Map every red_by to a Klass 1,2,...31,32,33,...
    int inklass_idx;                                    // idx of function inside its klass, ranging 0...31
    
    if(klass>32){
            klass = (klass-1) % 32 + 1; 
            inklass_idx = (red_by-1) % 32;   
    }else{        
        inklass_idx = red_by - ((klass-1)*32 + 1);              
    }     
     
    if(klass==32) klass=9;                              // Skip the merge-suspect 32rd klass!
     
    int o_idx, i_idx =inklass_idx, ab_idx;
    for(o_idx = 0; o_idx<6; o_idx++ ){
        ab_idx = in[i_idx]%64;
        out[o_idx] = alphabet[ ab_idx ];    
        i_idx = (i_idx + klass) % 32;
    }
   	out[o_idx] = '\0';									// NULL terminate it to process as C-string                     
}

		
		
/*
 * Reduction-functions' set #3.
 *          "Maximum output entropy"
 *
 * Resulting from some "mixing" of techniques
 *
 */
void red_functs_set3( char* out, const uint8_t *in, int red_by ){				

    int o_idx, i_idx =red_by%32, ab_idx;
    for(o_idx = 0; o_idx<6; o_idx++ ){
        ab_idx = in[i_idx]%64;
        out[o_idx] = alphabet[ ab_idx ];    
        i_idx = (i_idx + 9) % 32;
    }
   	out[o_idx] = '\0';									// NULL terminate it to process as C-string                     
}
		
		
		
		
/*
 * Reduction-functions' set #4.
 *          "PRNG - Uniformly distributed indices "
 *
 * Seeding uniform_distribution generator with red_by index
 * for the indices of the hash's bytes to be selected. 
 * Then these bytes' values are mapped with a mod64 to a 
 * character in the passwords alphabet
 *
 *  WARNING: Much slower to implement, but offers greater uniformity
 *
 */
void red_functs_set4( char* out, const uint8_t *in, int red_by ){				
    int o_idx, i_idx, ab_idx;
    int range_from  = 0;            // inclusive
    int range_to    = 31;
    default_random_engine           generator(red_by); // seeding happens here
    uniform_int_distribution<int>   distr(range_from, range_to);    
    
    for(o_idx = 0; o_idx<6; o_idx++ ){
        i_idx = distr(generator);
 
        ab_idx = in[i_idx]%64;
        out[o_idx] = alphabet[ ab_idx ];    
        i_idx = (i_idx + 9) % 32;
    }
   	out[o_idx] = '\0';									// NULL terminate it to process as C-string                     
}
		
	
	
	
/*
 * Reduction-functions' set #5.
 *          "PRNG - Uniformly distributed passchars "
 *
 * Red-by with a "twist" determines the byte of the hash
 * that will be selected to seed the PRNG.
 * This in turn generates 6 chars uniformly distributed in the alphabet
 * 
 */
void red_functs_set5( char* out, const uint8_t *in, int red_by ){				
    int o_idx, i_idx;
    int range_from  = 0;            // inclusive
    int range_to    = 63;
    
    i_idx = (red_by+9) % 32;
        
    default_random_engine           generator( in[i_idx] ); // seeding happens here
    uniform_int_distribution<int>   distr(range_from, range_to);    
    
    for(o_idx = 0; o_idx<6; o_idx++ ){
        out[o_idx] = alphabet[ distr(generator) ];    
    }
   	out[o_idx] = '\0';									// NULL terminate it to process as C-string                     
}	
		


void write_table( Table_t* rainbow,  ofstream& outtable ){
	
	outtable << "TABLE 1" << endl;
	
	for( pair<char[7],char[7]>* pp : *rainbow )
		outtable << pp->first << " -> " << pp->second << endl;
	
	outtable.put('\n');
}	




Table_t* read_table( ifstream& intable ){
	
	Table_t* rainbow = new Table_t();	
	char line[24];
	
	intable.getline(line, 9);									// Skip "TABLE #1\n"
	while(! intable.eof() ){
		intable.getline(line, 24);						
		if( strlen(line)<6 ) break;
		
		pair<char[7],char[7]>* pp = new pair<char[7],char[7]>();
		strncpy( pp->first, line, 6);
		pp->first[6] = '\0';
		
		strncpy( pp->second, line+10, 6);
		pp->second[6] = '\0';
		rainbow->push_back( pp );
	}			 	
	
	if(v){
		cout << "Table imported succesfully\nIt now contains: " 
			<< rainbow->size() << " chains" << endl;
	
//		ofstream out("whatwasread.txt", ofstream::trunc);
//		write_table(rainbow,out);
	}	
	
	return rainbow;
}	





/*
 * Seeker function to operate on sorted rainbow table.
 *
 * Given a target "pwd" password, a binary search is performed 
 * in endpoints for the first exact match.
 * 
 * If found it returns TRUE, and sets "startpoint" to point
 * at the beggining of the chain, ending in "pwd".
 *
 * If not found, FALSE is returned and "startpoint" is untouched.
 *
 */
bool lookup(char* pwd, Table_t* rainbow, char* startpoint){	
																		// Implicitly assigning ceil[ size/2 ] 
	int	i;
	int	base_index		          = 1;
	int	middle_index	          = rainbow->size() / 2;
	int	top_index		          = rainbow->size();		
	Table_t::iterator         it  = rainbow->begin();
	Table_t::reverse_iterator tit = rainbow->rbegin();	

	pair<char[7],char[7]>					*middle_pair;				// Pivot-pointer pair
	pair<char[7],char[7]>                   *base_pair = *it, *top_pair = *tit;
	pair<char[7],char[7]>					*dummy_pair = new pair<char[7],char[7]>();
	strcpy( dummy_pair->second, pwd );									
																		// Create dummy pair, with only an endpoint
																		//  that will be passed to comparator 
																																			
	for(i=base_index ; i!=middle_index ; i++,it++) ;					// Move to the middle of the initial-list
	

	while( true ){														// Split list to halves repeatedly	
		middle_pair = *it;												// Set the pivot-pointer 
		if( middle_index==base_index ||
			middle_index==top_index ) break;	 						//  and break loop for the final comparison 
																		//  if list is reduced to a single element
																		
	/*	if(v){
			sleep(3);
			cout << "[*it,base,middle,top] = [ "
				<< (*it)->second << ' ' << base_index << ' ' 
				<< middle_index << ' ' << top_index <<']' << endl;																
		}
	*/			
																		
		if( pass_comparator(dummy_pair,base_pair)                       // Skip the whole process in the rare case
		 || pass_comparator(top_pair,dummy_pair)  ){                    // that pwd > top OR pwd < base
		     delete dummy_pair;
		     return false;
		}     
		 
		 
		if(! strcmp(middle_pair->second,pwd) ){							// If pwd = middle, return results
			strcpy( startpoint, middle_pair->first);
			delete dummy_pair;
			return true;
		}
		else if ( pass_comparator(dummy_pair,middle_pair) ){			// If pwd < middle
		//	base_index = base_index;									// Update indexes
			top_index = middle_index;	
			middle_index = base_index + (top_index - base_index) / 2;							
			for( ; i!=middle_index ; i--,it-- )	;						// Drop pivot to new middle
			continue;													// Loop again
		}
		else{															// If pwd > middle, loop again for the bigger-ones half
			base_index = middle_index;									// Update indexes
		//	top_index = top_index;	
			middle_index = base_index + (top_index - base_index) / 2;			
			for(i=base_index ; i!=middle_index ; i++,it++) ;			// Raise pivot to new middle
			continue;													// Loop again
		}
	}
	
	delete dummy_pair;
	if(! strcmp(middle_pair->second,pwd) ){								// Make one last comparisonfor single point case
		strcpy( startpoint, middle_pair->first);
		return true;
	}
	else{
		return false;
	}		
}





/*
 * "Hash-Reduce cycle" function to reproduce chain.
 *
 * This function is called when a hit is found in the rainbow
 * table's endpoints for the given hash. 
 * 
 * It repeatedly switches the startpoint from the hash-domain
 * to the password-domain until a hash identical to the original one
 * is computed. The password that resulted to it is returned in the 
 * preallocated character buffer "password" for the user to try.
 *
 */
bool follow_chain( uint8_t hash[32], char* startpoint, char* password, int t){

	uint8_t	cur_h[32];
	char	cur_pwd[7];


	for(int i=1; i<t+2 ; i++){												
		if(i==1)	hashfun( cur_h,(uint8_t*)startpoint,6 );
		else		hashfun( cur_h,(uint8_t*)cur_pwd,6 );				
		
		if( memcmp(cur_h,hash,sizeof(cur_h))==0 ){
		    strcpy(password, cur_pwd);	
		    if(v)
		        cout<<"After applying red-fun #"<<i-1<<", the hashed "
		            <<"output matches the originating one!"<<endl;
		    return true;    
		}
		
		redfun( cur_pwd,cur_h,i );
	}
	return false;
}




/*
 * Success probability calculator,
 * based on the Oechslin formula for
 * standard rainbow tables
 *
 */
double success_prob(long m, int t, int l, long long N){
    double *mi = new double[t];
    
    mi[0] = m;
    for(int i=1 ; i<t ; i++)
        mi[i] = N * (1-exp( -mi[i-1]/N ));
        
    double prob = 1;
    for(int i=0 ; i<t ; i++)
        prob *= (1 - (mi[i]/N));  
        
    delete[] mi;      
        
    return (1 - pow(prob,(float)l)  ) ;    
}
















